prediction_market_agent/agents/microchain_agent/app.py

```diff
@@ -40,9 +30,48 @@ def run_agent(agent: Agent, iterations: int, model: str) -> None:
 def execute_reasoning(agent: Agent, reasoning: str, model: str) -> None:
     with openai_costs(model) as costs:
         agent.execute_command(f'Reasoning("{reasoning}")')
+        display_new_history_callback(agent)  # Run manually after `execute_command`
         st.session_state.running_cost += costs.cost
 
 
+def display_all_history(agent: Agent) -> None:
+    """
+    Display the agent's history in the Streamlit app.
+    """
+    # Skip the initial messages
+    history = agent.history[get_initial_history_length(agent) :]
+
+    for h in history:
+        st.chat_message(h["role"]).write(h["content"])
+
+
+def display_new_history_callback(agent: Agent) -> None:
+    """
+    A callback to display the agent's history in the Streamlit app after a run
+    with a single interation.
+    """
+    history_depth = 2  # One for the user input, one for the agent's reply
+    history = agent.history[-history_depth:]
+    for h in history:
+        st.chat_message(h["role"]).write(h["content"])
+
+
+def agent_is_initialized() -> bool:
+    return "agent" in st.session_state
+
+
+def maybe_initialize_agent(model: str) -> None:
+    # Initialize the agent
+    if not agent_is_initialized():
+        st.session_state.agent = get_agent(market_type=MarketType.OMEN, model=model)
+        st.session_state.agent.reset()
```

Comment:
This might be the wrong place, but I see a discrepancy between the model used and the model logged.

1. Model selected on streamlit

![image](https://github.com/gnosis/prediction-market-agent/assets/4126235/96b9c81c-93fc-450f-900f-0d6905ab5991)


2 - Model logged
![Uploading image.png…]()

Reply:
Ah yes, this `ENGINE: gpt-4...` logging comes from the local mech code. You can't universally pass in a model arg there, so I left it to default behaviour. Once https://github.com/gnosis/prediction-market-agent/pull/124 is merged this issue will go away